src/Accounts/Authentication/Utilities/FrequencyService.cs

```diff
@@ -0,0 +1,195 @@
+            }
+            else
+            {
+                throw new AzPSArgumentException ($"Feature name '{featureName}' not found in FrequencyService", nameof(featureName));
+            }
+
+        }
+        /// <summary>
+        /// This method is used for testing purposes only. It allows the caller to specify a fakeNow time to use for the check.
+        /// </summary>
+        internal void Check(string featureName, Func<bool> businessCheck, Action business, DateTime fakeNow)
+        {
+            if (_perPSSessionRegistry != null && _perPSSessionRegistry.ContainsKey(featureName))
+            {
+                if (_perPSSessionRegistry[featureName] == false && businessCheck())
+                {
+                    _perPSSessionRegistry[featureName] = true;
+                    business();
+                }
+            }
+            else if (_frequencies.ContainsKey(featureName))
+            {
+                if (_clock.IsDue(_frequencies[featureName].LastCheckTime, _frequencies[featureName].Frequency) && businessCheck())
+                {
+                    _frequencies[featureName].LastCheckTime = fakeNow;
+                    business();
+                    Save();
+                }
+            }
+            else
+            {
+                throw new AzPSArgumentException ($"Feature name '{featureName}' not found in FrequencyService", nameof(featureName));
+            }
+
+        }
+
+        public void Register(string featureName, TimeSpan frequency)
+        {
+            if (!_frequencies.ContainsKey(featureName))
+            {
+                _frequencies.Add(featureName, new FrequencyInfo(frequency, DateTime.MinValue));
+                Save();
+            }
+            if (_frequencies.ContainsKey(featureName) && _frequencies[featureName].Frequency != frequency)
+            {
+                throw new AzPSArgumentException ($"Feature name '{featureName}' already exists in FrequencyService with a different frequency!", nameof(featureName));
+            }
+        }
+
+        public void RegisterInSession(string featureName)
+        {
+            if (_perPSSessionRegistry == null)
+            {
+                _perPSSessionRegistry = new Dictionary<string, bool>();
+            }
+            if (!_perPSSessionRegistry.ContainsKey(featureName))
+            {
+                _perPSSessionRegistry.Add(featureName, false);
+            }
+        }
+
+        public void Save()
+        {
+            string json = JsonConvert.SerializeObject(_frequencies);
+            _dataStore.WriteFile(_filePath, json);
+        }
+
+        internal List<string> GetAllFeatureNames()
+        {
+            var allFeatureNames = new List<string>(_frequencies.Keys);
+            if (SessionLogic != null)
+            {
+                allFeatureNames.AddRange(SessionLogic.Keys);
+            }
+            return allFeatureNames.ToList();
+        }
+    }
+
+    public interface IClock
```

Comment:
Let's make this interface internal, in case we are going to modify it in the future.